### PR TITLE
crypto: Plugin API rework

### PR DESCRIPTION
**Pull request by vojtechtrefny**
This includes multiple changes to the plugin API:
- LUKS2 specific functions are removed.
- "_blob" functions that accept binary passphrase are removed.
- Functions that need a passphrase or key now expect a new
  internal struct BDCryptoKeyslotContext that users needs to
  initialize with passphrase or key file.

This is still very much work in progress, but feedback for the general idea is much welcomed.

Replaces #610

### CODE REVIEW

File: src/lib/plugin_apis/crypto.api [hunk 1]
```diff
@@ -703,86 +703,124 @@ guint64 bd_crypto_luks_get_metadata_size (const gchar *device, GError **error);
...
+            gsize data_len;
+        } passphrase;
+
+        struct {
+            gchar *keyfile;
+            guint64 keyfile_offset;
+            gsize key_size;
+        } keyfile;
+    } u;
+} BDCryptoKeyslotContext;
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/lib/plugin_apis/crypto.api:729]: That's what I was afraid of!

**Comment by vojtechtrefny** [comment]
There are few things missing like documentation but the code should be finished now and mostly ready for review.

File: src/lib/plugin_apis/crypto.api [hunk 2]
```diff
@@ -376,6 +377,7 @@ typedef struct BDCryptoLUKSInfo {
     gchar *uuid;
     gchar *backing_device;
     gint64 sector_size;
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/lib/plugin_apis/crypto.api:379]: If you have a look at `crypto.h` and search for `sector_size`, you could see data types of `guint64`, `guint32` and `gint64`, although in different context. Range of return values of `crypt_get_sector_size()` is undocumented, reading the source code wouldn't give me much clue either.

File: src/lib/plugin_apis/crypto.api (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/lib/plugin_apis/crypto.api:729]: This seems to work for me, tests are passing:

```
struct _BDCryptoKeyslotContext {
...
};
typedef struct _BDCryptoKeyslotContext BDCryptoKeyslotContext;
```

**Review by tbzatek — APPROVED** [review]
Apart from those two open issues the rest looks good

File: src/lib/plugin_apis/crypto.api (hunk 2 above)
[the remark this answers is not in the document]
Comment (vojtechtrefny) [reply src/lib/plugin_apis/crypto.api]: Ok, I changed everything to `guint32`. Not sure if it's the best type, but at least now every sector size should be the same type.

File: src/lib/plugin_apis/crypto.api (hunk 1 above)
[the remark this answers is not in the document]
Comment (vojtechtrefny) [reply src/lib/plugin_apis/crypto.api]: Ok, seems to be working fine.

File: src/lib/plugin_apis/crypto.api (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/lib/plugin_apis/crypto.api]: Looks fine, thanks.

File: src/plugins/crypto.c [hunk 3]
```diff
@@ -1433,196 +1327,138 @@ gboolean bd_crypto_luks_add_key (const gchar *device, const gchar *pass, const g
...
-gboolean bd_crypto_luks_remove_key (const gchar *device, const gchar *pass, const gchar *key_file, GError **error) {
+gboolean bd_crypto_luks_remove_key (const gchar *device, BDCryptoKeyslotContext *context, GError **error) {
     struct crypt_device *cd = NULL;
     gint ret = 0;
     guint64 progress_id = 0;
     gchar *msg = NULL;
-    gboolean success = FALSE;
     gchar *key_buf = NULL;
     gsize buf_len = 0;
+    GError *l_error = NULL;
```
Comment (vpodzime) [inline src/plugins/crypto.c:1410]: Why this change to use a local error variable (only) in this function?

File: src/python/gi/overrides/BlockDev.py [hunk 4]
```diff
@@ -262,7 +262,7 @@ def __init__(self, *args, **kwargs):   # pylint: disable=unused-argument
 __all__.append("CryptoLUKSExtra")
 
 class CryptoKeyslotContext(BlockDev.CryptoKeyslotContext):
-    def __new__(cls, passphrase=None, keyfile=None, keyfile_offset=0, key_size=0):
+    def __new__(cls, passphrase=None, keyfile=None, keyfile_offset=0, key_size=0, keyring=None):
         if (passphrase and keyfile):
```
Comment (vpodzime) [inline src/python/gi/overrides/BlockDev.py:266]: should now check `keyring` here too, right?

File: src/lib/plugin_apis/crypto.api [hunk 5]
```diff
@@ -1068,10 +1094,9 @@ BDCryptoLUKSTokenInfo** bd_crypto_luks_token_info (const gchar *device, GError *
 /**
  * bd_crypto_integrity_format:
  * @device: a device to format as integrity
- * @algorithm: integrity algorithm specification (e.g. "crc32c" or "sha256") or %NULL to use the default
+ * @algorithm: integrity algorithm specification (e.g. "crc32c" or "sha256")
```
Comment (vpodzime) [inline src/lib/plugin_apis/crypto.api:1120]: I don't see any change of how `algorithm` is handled (in this commit).

**Review by vpodzime — APPROVED** [review]
+ *passphrase/~kye~key* in one of the commit messages.

Looks good to me otherwise. GOOD JOB! :clap: :clap::clap::clap::clap::clap:

File: src/python/gi/overrides/BlockDev.py (hunk 4 above)
Comment (vojtechtrefny) [reply src/python/gi/overrides/BlockDev.py]: Fixed.

File: src/plugins/crypto.c (hunk 3 above)
Comment (vojtechtrefny) [reply src/plugins/crypto.c:1410]: Moved to a separate commit.

File: src/lib/plugin_apis/crypto.api (hunk 5 above)
Comment (vojtechtrefny) [reply src/lib/plugin_apis/crypto.api:1120]: That was probably some copy-pasting of docstrings between the C and API files. But we shouldn't allow NULL anyway, I added an extra commit dealing with this.
